src/sqlfluff/rules/L052.py

```diff
@@ -224,27 +218,54 @@ def _handle_semicolon_newline(
             )
         else:
             fixes.extend(
-                [
-                    LintFix.create_after(
-                        self._choose_anchor_segment(
-                            context, "create_after", anchor_segment, filter_meta=True
-                        ),
-                        [
-                            NewlineSegment(),
-                            SymbolSegment(raw=";", type="symbol", name="semicolon"),
-                        ],
-                    ),
-                    LintFix.delete(
-                        context.segment,
-                    ),
-                ]
+                self._create_semicolon_and_delete_whitespace(
+                    context,
+                    anchor_segment,
+                    info.whitespace_deletions,
+                    [
+                        NewlineSegment(),
+                        SymbolSegment(raw=";", type="symbol", name="semicolon"),
+                    ],
+                )
             )
-            fixes.extend(LintFix.delete(d) for d in info.whitespace_deletions)
         return LintResult(
             anchor=anchor_segment,
             fixes=fixes,
         )
 
+    def _create_semicolon_and_delete_whitespace(
+        self,
+        context: RuleContext,
+        anchor_segment: BaseSegment,
+        whitespace_deletions: Segments,
+        create_segments: List[BaseSegment],
+    ) -> List[LintFix]:
+        anchor_segment = self._choose_anchor_segment(
+            context, "create_after", anchor_segment, filter_meta=True
+        )
+        lintfix_fn = LintFix.create_after
+        # :TRICKY: Use IdentitySet rather than set() since
+        # different segments may compare as equal.
+        whitespace_deletion_set = IdentitySet(whitespace_deletions)
+        if anchor_segment in whitespace_deletion_set:
+            # Can't delete() and create_after() the same segment. Use replace()
+            # instead.
```

Comment:
This comment feels more like an explanation for this PR, rather than something that needs to be in the code.

```suggestion
```

Reply:
Really? I think it's helpful -- it explains in the line below why we are switching `lintfix_fn` from `create_after` to `replace`, which otherwise may not be obvious to a future maintainer. Do you think there's a better way to explain what's happening here?

Reply:
Ok see that now. Thought it was an explanation of why you rewrote it the way you did. Resolving